src/middleware.rs

```diff
@@ -447,133 +480,172 @@ pub enum NativeOperation {
-        }
-    }
-    pub fn key(&self) -> Result<AnchoredKey> {
+impl Operation {
+    pub fn code(&self) -> NativeOperation {
+        use NativeOperation::*;
         match self {
-            Self::Key(ak) => Ok(ak.clone()),
-            _ => Err(anyhow!("Operation argument {:?} is not a key.", self)),
+            Self::None => None,
+            Self::NewEntry => NewEntry,
+            Self::CopyStatement(_) => CopyStatement,
+            Self::EqualFromEntries(_, _) => EqualFromEntries,
+            Self::NotEqualFromEntries(_, _) => NotEqualFromEntries,
+            Self::GtFromEntries(_, _) => GtFromEntries,
+            Self::LtFromEntries(_, _) => LtFromEntries,
+            Self::TransitiveEqualFromStatements(_, _) => TransitiveEqualFromStatements,
+            Self::GtToNotEqual(_) => GtToNotEqual,
+            Self::LtToNotEqual(_) => LtToNotEqual,
+            Self::ContainsFromEntries(_, _) => ContainsFromEntries,
+            Self::NotContainsFromEntries(_, _) => NotContainsFromEntries,
+            Self::RenameContainedBy(_, _) => RenameContainedBy,
+            Self::SumOf(_, _, _) => SumOf,
+            Self::ProductOf(_, _, _) => ProductOf,
+            Self::MaxOf(_, _, _) => MaxOf,
         }
     }
-}
 
-// TODO: Replace this with a more stringly typed enum as in the Devcon implementation.
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub struct Operation(pub NativeOperation, pub Vec<OperationArg>);
-
-impl Operation {
-    pub fn code(&self) -> NativeOperation {
-        self.0
+    pub fn args(&self) -> Vec<Statement> {
+        match self.clone() {
+            Self::None => vec![],
+            Self::NewEntry => vec![],
+            Self::CopyStatement(s) => vec![s],
+            Self::EqualFromEntries(s1, s2) => vec![s1, s2],
+            Self::NotEqualFromEntries(s1, s2) => vec![s1, s2],
+            Self::GtFromEntries(s1, s2) => vec![s1, s2],
+            Self::LtFromEntries(s1, s2) => vec![s1, s2],
+            Self::TransitiveEqualFromStatements(s1, s2) => vec![s1, s2],
+            Self::GtToNotEqual(s) => vec![s],
+            Self::LtToNotEqual(s) => vec![s],
+            Self::ContainsFromEntries(s1, s2) => vec![s1, s2],
+            Self::NotContainsFromEntries(s1, s2) => vec![s1, s2],
+            Self::RenameContainedBy(s1, s2) => vec![s1, s2],
+            Self::SumOf(s1, s2, s3) => vec![s1, s2, s3],
+            Self::ProductOf(s1, s2, s3) => vec![s1, s2, s3],
+            Self::MaxOf(s1, s2, s3) => vec![s1, s2, s3],
+        }
     }
-    pub fn args(&self) -> &[OperationArg] {
-        &self.1
+    /// Forms operation from op-code and arguments.
+    pub fn op(op_code: NativeOperation, args: &[Statement]) -> Result<Self> {
+        type NO = NativeOperation;
+        let arg_tup = (
+            args.get(0).cloned(),
+            args.get(1).cloned(),
+            args.get(2).cloned(),
+        );
+        Ok(match (op_code, arg_tup) {
```

Comment:
Maybe we also need to include `args.len()` in the match to validate the arguments length of each operation